### PR TITLE
CGAL: Add remesh_surface utility

### PR DESCRIPTION
**Pull request by fdrmrc**
As the name says, this utility allows to remesh CGAL surfaces. The main purpose of this is to improve the quality of grids coming from boolean operations. As an example, one can start from [doc/doxygen/images/boolean_union_hyper_spheres.png](https://github.com/dealii/dealii/compare/master...luca-heltai:cgal-Surface_remeshing?expand=1#diff-0f91bdb8208b856d034251875a4e0d4ffdd6e35a4c7f1c606730f5d9551060dc) and get [doc/doxygen/images/boolean_union_hyper_spheres_remeshed.png](https://github.com/dealii/dealii/compare/master...luca-heltai:cgal-Surface_remeshing?expand=1#diff-fe2f4757336eeac26e9c7b8d7338c19d4908ac4269c22449d6416ea03600e7cb)

### CODE REVIEW

File: include/deal.II/cgal/utilities.h [hunk 1]
```diff
@@ -198,6 +203,37 @@ namespace CGALWrappers
     const CGAL::Surface_mesh<CGALPointType> &surface_mesh_2,
     const BooleanOperation &                 boolean_operation,
     CGAL::Surface_mesh<CGALPointType> &      output_surface_mesh);
+
+  /**
+   * Remesh a CGAL::Surface_mesh according to a set of given named parameters.
+   * See more about them at:
+   * https://doc.cgal.org/latest/Mesh_3/classCGAL_1_1Mesh__criteria__3.html.
+   *
```
Comment (luca-heltai) [inline include/deal.II/cgal/utilities.h:211]: Can you copy here the blob about the parameters that we added to the other functions? We don't want users to go elsewhere for the documentation if it is possible to have it right here...

File: tests/cgal/cgal_surface_mesh_05.cc [hunk 2]
```diff
@@ -70,10 +70,9 @@ test()
                             out_mesh);
   // Now back to deal.II
   cgal_surface_mesh_to_dealii_triangulation(out_mesh, tria_out);
-  std::ofstream out_name_spheres("boolean_triangulation_spheres.vtk");
+  std::ofstream out_name_spheres("boolean_intersection_hyper_spheres.vtk");
   go.write_vtk(tria_out, out_name_spheres);
   deallog << "OK" << std::endl;
-  remove("tria_out.vtk");
```
Comment (luca-heltai) [inline tests/cgal/cgal_surface_mesh_05.cc:76]: can you call `remove("boolean_intersection_hyper_spheres.vtk")`?

File: tests/cgal/cgal_surface_mesh_05.cc [hunk 3]
```diff
@@ -106,9 +105,8 @@ test()
                             out_mesh);
   // Now back to deal.II
   cgal_surface_mesh_to_dealii_triangulation(out_mesh, tria_out);
-  std::ofstream out_name_cubes("boolean_triangulation_cubes.vtk");
+  std::ofstream out_name_cubes("boolean_intersection_cubes.vtk");
   go.write_vtk(tria_out, out_name_cubes);
-  remove("tria_out.vtk");
   deallog << "OK" << std::endl;
```
Comment (luca-heltai) [inline tests/cgal/cgal_surface_mesh_05.cc:110]: same here. `remove("boolean_intersection_cubes.vtk");`

**Review by luca-heltai — CHANGES_REQUESTED** [review]

**Comment by drwells** [comment]
We are at the point where we need to focus on fixing bugs and not adding new features for 9.4. Can we push this back to 9.5?

**Comment by fdrmrc** [comment]
I think this one is pretty safe, as the body of this function is taken verbatim from a CGAL example program. But of course I am totally fine with postponing this if you think it's better @drwells

**Review by luca-heltai — APPROVED** [review]

File: include/deal.II/cgal/triangulation.h [hunk 4]
```diff
@@ -181,6 +182,84 @@ namespace CGALWrappers
...
+   * generated Triangulation object
+   */
+  template <int dim>
+  void
+  implicit_function(Triangulation<dim, 3> &                  tria,
+                    const Function<3> &                      implicit_function,
+                    const CGALWrappers::AdditionalData<dim> &data =
+                      CGALWrappers::AdditionalData<dim>{},
+                    const Point<3> &interior_point    = Point<3>(),
+                    const double &  outer_ball_radius = 1.0);
```
Comment (kronbichler) [inline include/deal.II/cgal/triangulation.h:241]: How is this function related to the intent of this PR? I also cannot find the definition of the function, so I assume this is spurious? I guess we should aim to calm down in adding new features from CGAL at this stage of the release cycle.

Comment (fdrmrc) [reply include/deal.II/cgal/triangulation.h]: I'm really sorry, I lost it while rebasing. Haste makes waste. Now it's fixed.
According to [CGALSupport](https://github.com/dealii/dealii/projects/25) this is the last CGAL-related PR for 9.4, but I'm fine with postponing it, as the last thing I want is to introduce bugs.

**Comment by tamiko** [comment]
@luca-heltai @kronbichler I need a call on this pull request regarding the release :slightly_smiling_face:

**Review by kronbichler — APPROVED** [review]
This looks good me now.
